packages/eslint-plugin/src/rules/no-duplicate-type-union-intersection-members.ts

```diff
@@ -0,0 +1,122 @@
+import type { TSESLint, TSESTree } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+
+import * as util from '../util';
+
+export type Options = [
+  {
+    checkIntersections?: boolean;
+    checkUnions?: boolean;
+  },
+];
+export type MessageIds = 'duplicate' | 'suggestFix';
+
+export default util.createRule<Options, MessageIds>({
+  name: 'no-duplicate-type-union-intersection-members',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'Disallow duplicate union/intersection type members',
+      recommended: false,
+    },
+    fixable: 'code',
+    hasSuggestions: true,
+    messages: {
+      duplicate: '{{type}} type member {{name}} is duplicated.',
+      suggestFix: 'Delete duplicated members of type (removes all comments).',
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          checkIntersections: {
+            type: 'boolean',
+          },
+          checkUnions: {
+            type: 'boolean',
+          },
+        },
+      },
+    ],
+  },
+  defaultOptions: [
+    {
+      checkIntersections: true,
+      checkUnions: true,
+    },
+  ],
+  create(context, [{ checkIntersections, checkUnions }]) {
+    const sourceCode = context.getSourceCode();
+    function checkDuplicate(
+      node: TSESTree.TSIntersectionType | TSESTree.TSUnionType,
+    ): void {
+      const duplicateMemberTexts: string[] = [];
+      const uniqueMemberTexts = new Set<string>();
+
+      const source = node.types.map(type => {
+        return {
+          node: type,
+          text: sourceCode.getText(type),
```

Comment:
We generally try not to use `sourceCode.getText` or similar in rules if at all possible. It's less reliable than comparing AST and/or types - for example, it'd be very confusing to users why adding a comment inside a type stops this rule from complaining on it. I'm not looking forward to fielding those bug reports 😄.

Does using the type checker and checking types for `===` work? I.e. building a `Set` of types for each union/intersection, and complaining for each type that's the same as a previous one?

Obligatory grumble that https://github.com/microsoft/TypeScript/issues/9879 isn't implemented...